src/NodeServices/PythonServices.cs

```diff
@@ -0,0 +1,241 @@
+        /// </summary>
+        private PythonEngineManager()
+        {
+            AvailableEngines = new ObservableCollection<PythonEngine>();
+
+            ScanPythonEngines();
+            AppDomain.CurrentDomain.AssemblyLoad += new AssemblyLoadEventHandler(AssemblyLoadEventHandler);
+        }
+
+        internal PythonEngine GetEngine(string version)
+        {
+            return AvailableEngines.FirstOrDefault(x => x.Name == version);
+        }
+
+        /// <summary>
+        /// Scan loaded Python engines
+        /// </summary>
+        internal void ScanPythonEngines()
+        {
+            var assems = AppDomain.CurrentDomain.GetAssemblies();
+            foreach (var assembly in assems)
+            {
+                ScanPythonEngine(assembly);
+            }
+        }
+
+        /// <summary>
+        /// The shared logic for Python node evaluation
+        /// </summary>
+        /// <param name="engine"></param>
+        /// <param name="evaluatorClass"></param>
+        /// <param name="evaluationMethod"></param>
+        internal void GetEvaluatorInfo(PythonEngineVersion engine, out string evaluatorClass, out string evaluationMethod)
+        {
+            // Provide evaluator info when the selected engine is loaded
+            if (engine == PythonEngineVersion.IronPython2 &&
+                AvailableEngines.Any(x => x.Version == PythonEngineVersion.IronPython2))
+            {
+                evaluatorClass = IronPythonEvaluatorClass;
+                evaluationMethod = IronPythonEvaluationMethod;
+                return;
+            }
+            if (engine == PythonEngineVersion.CPython3 &&
+                AvailableEngines.Any(x => x.Version == PythonEngineVersion.CPython3))
+            {
+                evaluatorClass = CPythonEvaluatorClass;
+                evaluationMethod = CPythonEvaluationMethod;
+                return;
+            }
+
+            // Throwing at the compilation stage is handled as a non-retryable error by the Dynamo engine.
+            // Instead, we want to produce an error at the evaluation stage, so we can eventually recover.
+            // We handle this by providing a dummy Python evaluator that will evaluate to an error message.
+            evaluatorClass = DummyEvaluatorClass;
+            evaluationMethod = DummyEvaluatorMethod;
+        }
+
+        private void AssemblyLoadEventHandler(object sender, AssemblyLoadEventArgs args)
+        {
+            PythonEngineManager.Instance.ScanPythonEngine(args.LoadedAssembly);
+        }
+
+        private void ScanPythonEngine(Assembly assembly)
+        {
+            if (assembly == null)
+            {
+                return;
+            }
+
+            PythonEngine engine = null;
+            // Currently we are using try-catch to validate loaded assembly and evaluation method exist
+            // but we can optimize by checking all loaded types against evaluators interface later
```

Comment:
do you mean interface or abstract base class?

Reply:
abstract base class..will change